lib/_stream_wrap.js

```diff
@@ -111,10 +138,81 @@ StreamWrap.prototype.write = function write(req, bufs) {
           errCode = uv.UV_EPIPE;
       }
 
-      self._handle.doAfterWrite(req);
-      self._handle.finishWrite(req, errCode);
+      handle.doAfterWrite(req);
+      handle.finishWrite(req, errCode);
     });
   }
 
   return 0;
 };
+
+function QueueItem(type, req) {
+  this.type = type;
+  this.req = req;
+  this.prev = this;
+  this.next = this;
+}
+
+StreamWrap.prototype._enqueue = function enqueue(type, req) {
+  const item = new QueueItem(type, req);
+  if (this._queue === null) {
+    this._queue = item;
+    return item;
+  }
+
+  item.next = this._queue.next;
+  item.prev = this._queue;
+  item.next.prev = item;
+  item.prev.next = item;
+
+  return item;
+};
+
+StreamWrap.prototype._dequeue = function dequeue(item) {
+  var next = item.next;
+  var prev = item.prev;
```

Comment:
This is when I wish we allowed macros. A simple `ASSERT(item instanceof QueueItem);` would be awesome. Oh well.